resources/assets/lib/stores/notification-stack-store.ts

```diff
@@ -0,0 +1,183 @@
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import DispatcherAction from 'actions/dispatcher-action';
+import { dispatchListener } from 'app-dispatcher';
+import DispatchListener from 'dispatch-listener';
+import NotificationJson, { NotificationBundleJson, NotificationStackJson, NotificationTypeJson } from 'interfaces/notification-json';
+import { action, observable } from 'mobx';
+import LegacyPmNotification from 'models/legacy-pm-notification';
+import Notification from 'models/notification';
+import NotificationStack, { idFromJson } from 'models/notification-stack';
+import NotificationType, { Name as NotificationTypeName  } from 'models/notification-type';
+import { nameToCategory } from 'notification-maps/category';
+import { NotificationEventMoreLoaded, NotificationEventNew } from 'notifications/notification-events';
+import { NotificationIdentity, resolveStackId } from 'notifications/notification-identity';
+import { NotificationResolver } from 'notifications/notification-resolver';
+import NotificationStore from './notification-store';
+
+@dispatchListener
+export default class NotificationStackStore implements DispatchListener {
+  @observable readonly stacks = new Map<string, NotificationStack>();
+  @observable readonly types = new Map<string | null, NotificationType>();
+  private readonly resolver = new NotificationResolver();
+
+  constructor(protected notificationStore: NotificationStore) {
+    this.addLegacyPm();
+  }
+
+  @action
+  flushStore() {
+    this.stacks.clear();
+    this.types.clear();
+    this.addLegacyPm();
+  }
+
+  getOrCreateType(identity: NotificationIdentity) {
+    let type = this.types.get(identity.objectType);
+    if (type == null) {
+      type = new NotificationType(identity.objectType, this.resolver);
+      this.types.set(type.name, type);
+    }
+
+    return type;
+  }
+
+  getStack(identity: NotificationIdentity) {
+    return this.stacks.get(resolveStackId(identity));
+  }
+
+  @action
+  handleDispatchAction(dispatched: DispatcherAction) {
+    if (dispatched instanceof NotificationEventNew) {
+      this.handleNotificationEventNew(dispatched);
+    } else if (dispatched instanceof NotificationEventMoreLoaded) {
+      this.handleNotificationEventMoreLoaded(dispatched);
+    }
+  }
```

Comment:
looking at adding a decorator that calls automatically registers `handle${objectType}` for this but it won't be done anytime soon